blockchain-importer/src/Pos/BlockchainImporter/Txp/Toil/Logic.hs

```diff
@@ -223,3 +249,38 @@ getBalanceUpdate txAux txUndo =
     let minusBalance = map (view _TxOut . toaOut) $ catMaybes $ toList txUndo
         plusBalance = map (view _TxOut) $ toList $ _txOutputs (taTx txAux)
     in BalanceUpdate {..}
+
+-- Returns the UxtoModifier corresponding to applying a list of txs
+applyUTxOModifier :: [(TxAux, TxUndo)] -> Txp.UtxoModifier
+applyUTxOModifier txs = mconcat $ applySingleModifier <$> txs
+
+-- Returns the UxtoModifier corresponding to applying a single tx
+applySingleModifier :: (TxAux, TxUndo) -> Txp.UtxoModifier
+applySingleModifier (txAux, _) = foldr  MM.delete
+                                        (foldr (uncurry MM.insert) mempty toInsert)
+                                        toDelete
+  where tx       = taTx txAux
+        id       = hash $ tx
+        outputs  = toList $ _txOutputs $ tx
+        toInsert = zipWith (\o index -> (TxInUtxo id index, TxOutAux o)) outputs [0..]
+        toDelete = toList $ _txInputs $ tx
+
+-- Returns the UxtoModifier corresponding to rollbacking a list of txs
+rollbackUTxOModifier :: [(TxAux, TxUndo)] -> Txp.UtxoModifier
+rollbackUTxOModifier txs = mconcat $ rollbackSingleModifier <$> reverse txs
+
+-- Returns the UxtoModifier corresponding to rollbacking a single tx
+rollbackSingleModifier :: (TxAux, TxUndo) -> Txp.UtxoModifier
+rollbackSingleModifier (txAux, txUndo) = foldr  MM.delete
+                                                (foldr (uncurry MM.insert) mempty toInsert)
+                                                toDelete
+  where tx       = taTx txAux
+        id       = hash $ tx
+        inputs   = toList $ _txInputs $ tx
+        outputs  = toList $ _txOutputs $ tx
+        toDelete = [ TxInUtxo id (fromIntegral index) | index <- [0..(length outputs - 1)] ]
+        toInsert = catMaybes $ zipWith mapValueToMaybe inputs $ toList txUndo
+
+        mapValueToMaybe :: a -> Maybe b -> Maybe (a, b)
```

Comment:
Pedantic detail: `mapValueToMaybe` could be defined as `\a -> fmap ((,) a)` (or even as `\a -> fmap (a,)` if `TupleSections` is enabled).